File: modules/background-timer/ios/BackgroundTimerModule.swift
import ExpoModulesCore

let BACKGROUND_TIMER_ACTIVE_EVENT_NAME = "onBackgroundTimeActive"

public class BackgroundTimerModule: Module {
  var backgroundtime = Double.infinity
  public func definition() -> ModuleDefinition {
    Name("BackgroundTimer")

    Events(BACKGROUND_TIMER_ACTIVE_EVENT_NAME)

    OnAppBecomesActive {
      let now = Date().timeIntervalSince1970 * 1000;
      let delta = now - self.backgroundtime
      self.sendEvent(BACKGROUND_TIMER_ACTIVE_EVENT_NAME,["delta": delta == -Double.infinity ? 0 : delta])
      self.backgroundtime = Double.infinity
    }
    
    OnAppEntersBackground {
      self.backgroundtime = Date().timeIntervalSince1970 * 1000;
    }
  }
}

File: modules/background-timer/android/src/main/java/expo/modules/backgroundtimer/BackgroundTimerModule.kt
package expo.modules.backgroundtimer

import androidx.core.os.bundleOf
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

const val BACKGROUND_TIMER_ACTIVE_EVENT_NAME = "onBackgroundTimeActive"

class BackgroundTimerModule : Module() {
  var backgroundtime = 0L
  override fun definition() = ModuleDefinition {
    Name("BackgroundTimer")

    Events(BACKGROUND_TIMER_ACTIVE_EVENT_NAME)

    OnActivityEntersForeground{
      val now = System.currentTimeMillis()
      val delta = now - backgroundtime
      sendEvent(BACKGROUND_TIMER_ACTIVE_EVENT_NAME, bundleOf("delta" to delta))
    }

    OnActivityEntersBackground{
      backgroundtime = System.currentTimeMillis()
    }
  }
}
